libsql-wal/src/registry.rs

```diff
@@ -0,0 +1,212 @@
+use crate::segment::{current::CurrentSegment, sealed::SealedSegment};
+use crate::shared_wal::SharedWal;
+use crate::transaction::{Transaction, WriteTransaction};
+
+/// Translates a path to a namespace name
+pub trait NamespaceResolver {
+    fn resolve(&self, path: &Path) -> NamespaceName;
+}
+
+impl<F: Fn(&Path) -> NamespaceName + Send + Sync + 'static> NamespaceResolver for F {
+    fn resolve(&self, path: &Path) -> NamespaceName {
+        (self)(path)
+    }
+}
+
+/// Wal Registry maintains a set of shared Wal, and their respective set of files.
+pub struct WalRegistry<FS: FileSystem> {
+    fs: FS,
+    path: PathBuf,
+    shutdown: AtomicBool,
+    opened: RwLock<HashMap<NamespaceName, Arc<SharedWal<FS>>>>,
+    resolver: Box<dyn NamespaceResolver + Send + Sync + 'static>,
+}
+
+impl WalRegistry<StdFs> {
+    pub fn new(
+        path: PathBuf,
+        resolver: impl NamespaceResolver + Send + Sync + 'static,
+    ) -> Result<Self> {
+        Self::new_with_fs(StdFs(()), path, resolver)
+    }
+}
+
+impl<FS: FileSystem> WalRegistry<FS> {
+    pub fn new_with_fs(
+        fs: FS,
+        path: PathBuf,
+        resolver: impl NamespaceResolver + Send + Sync + 'static,
+    ) -> Result<Self> {
+        fs.create_dir_all(&path)?;
+        Ok(Self {
+            fs,
+            path,
+            opened: Default::default(),
+            shutdown: Default::default(),
+            resolver: Box::new(resolver),
+        })
+    }
+
+    #[tracing::instrument(skip(self, db_path))]
+    pub fn open(self: Arc<Self>, db_path: &Path) -> Result<Arc<SharedWal<FS>>> {
+        if self.shutdown.load(Ordering::SeqCst) {
+            todo!("open after shutdown");
+        }
+
+        let namespace = self.resolver.resolve(db_path);
+        let mut opened = self.opened.upgradable_read();
+        if let Some(entry) = opened.get(&namespace) {
+            return Ok(entry.clone());
+        }
+
+        let path = self.path.join(namespace.as_str());
+        self.fs.create_dir_all(&path)?;
+        let dir = walkdir::WalkDir::new(&path).sort_by_file_name().into_iter();
+
+        let tail = SegmentList::default();
+        for entry in dir {
+            let entry = entry.map_err(|e| e.into_io_error().unwrap())?;
+            if entry
+                .path()
+                .extension()
+                .map(|e| e.to_str().unwrap() != "seg")
+                .unwrap_or(true)
+            {
+                continue;
+            }
+
+            let file = self.fs.open(false, true, true, entry.path())?;
+
+            if let Some(sealed) =
+                SealedSegment::open(file.into(), entry.path().to_path_buf(), Default::default())?
+            {
+                tail.push_log(sealed);
+            }
+        }
+
+        let db_file = self.fs.open(false, true, true, db_path)?;
+
+        // If this is a fresh database, we want to patch the header value for reserved space at the
+        // end of the file to store the replication index
```

Comment:
btw we need to confirm that updating this header field is okay with the encryption schemes we use. 

for instance [SQLite Checksum](https://www.sqlite.org/cksumvfs.html) uses reserved bytes and changing it breaks the checksum logic. I was wondering if this is used in encryption too.

Reply:
yeah actually this comment is stale, we never do that. I think the next iteration of encryption should be done at the wal level though